GLPT_Cocoa.inc

```diff
@@ -1,89 +1,390 @@
-  //  Include(Result, ssAlt);
-  //if (State[VK_CONTROL] and 128) <> 0 then
-  //  Include(Result, ssCtrl);
-  //if (State[VK_LBUTTON] and 128) <> 0 then
-  //  Include(Result, ssLeft);
-  //if (State[VK_RBUTTON] and 128) <> 0 then
-  //  Include(Result, ssRight);
-  //if (State[VK_MBUTTON] and 128) <> 0 then
-  //  Include(Result, ssMiddle);
-  //if (State[VK_CAPITAL] and 1) <> 0 then
-  //  Include(Result, ssCaps);
-  //if (State[VK_NUMLOCK] and 1) <> 0 then
-  //  Include(Result, ssNum);
-  //if (State[VK_SCROLL] and 1) <> 0 then
-  //  Include(Result, ssScroll);
+
+  if modifierFlags = (modifierFlags or NSShiftKeyMask) then
+    Include(Result, ssShift);
+
+  if modifierFlags = (modifierFlags or NSControlKeyMask) then
+    Include(Result, ssCtrl);
+
+  if modifierFlags = (modifierFlags or NSAlternateKeyMask) then
+    Include(Result, ssAlt);
+
+  if modifierFlags = (modifierFlags or NSCommandKeyMask) then
+    Include(Result, ssSuper);
+end;
+
+
+procedure SetupMainMenu;
+
+  procedure AddMenu (menu: NSMenu);
+  var
+    menuItem: NSMenuItem;
+  begin
+    menuItem := NSMenuItem.alloc.initWithTitle_action_keyEquivalent(menu.title, nil, NSSTR('')).autorelease;
+    menuItem.setSubmenu(menu);
+    NSApp.mainMenu.addItem(menuItem);
+  end;
+
+var
+  mainMenu: NSMenu;
+  appleMenu: NSMenu;
+  windowMenu: NSMenu;
+  menuItem: NSMenuItem;
+begin
+
+  // main menu
+  mainMenu := NSMenu.alloc.init.autorelease;
+  NSApp.setMainMenu(mainMenu);
+
+  // apple menu
+  appleMenu := NSMenu.alloc.initWithTitle(NSSTR('')).autorelease;
+  appleMenu.addItemWithTitle_action_keyEquivalent(NSSTR('Quit'), objcselector('terminate:'), NSSTR('q'));
+
+  AddMenu(appleMenu);
+
+  // window menu
+  windowMenu := NSMenu.alloc.initWithTitle(NSSTR('Window')).autorelease;
+  windowMenu.addItemWithTitle_action_keyEquivalent(NSSTR('Minimize'), objcselector('performMiniaturize:'), NSSTR('m'));
+  windowMenu.addItemWithTitle_action_keyEquivalent(NSSTR('Zoom'), objcselector('performZoom:'), NSSTR(''));
+
+  AddMenu(windowMenu);
+
 end;
 
 function Cocoa_Init: boolean;
+var
+  pool: NSAutoreleasePool;
+  app: TCocoaApp;
 begin
-  {$note implement this}
+  // https://cocoawithlove.com/2009/01/demystifying-nsapplication-by.html
+  
+  pool := NSAutoreleasePool.alloc.init;
+
+  app := TCocoaApp(TCocoaApp.sharedApplication);
+
+  if NSApp.mainMenu = nil then
+    SetupMainMenu;
+  app.finishLaunching;
+
+  NSApp.activateIgnoringOtherApps(true);
+
+  pool.release;
 
   exit(True);
 end;
 
 procedure Cocoa_SetCursor(cursor: byte);
 begin
-  {$note implement this}
+  case cursor of
+    GLPT_ARROW_CURSOR:
+      NSCursor.arrowCursor.set_;
+    GLPT_IBEAM_CURSOR:
+      NSCursor.IBeamCursor.set_;
+    GLPT_CROSSHAIR_CURSOR:
+      NSCursor.crosshairCursor.set_;
+    GLPT_HAND_CURSOR:
+      NSCursor.openHandCursor.set_;
+    GLPT_HRESIZE_CURSOR:
+      NSCursor.resizeLeftRightCursor.set_;
+    GLPT_VRESIZE_CURSOR:
+      NSCursor.resizeUpDownCursor.set_;
+    otherwise
+      halt(-1);
```

Comment:
I would use the default arrow cursor instead of halting the application. Possibly initiate a warning message at the same time.